src/main/c/NativeModule.cpp

```diff
@@ -0,0 +1,559 @@
+    fprintf(f, "  -I$(JAVA_INCLUDE) \\\n");
+    fprintf(f, "  -I$(JAVA_INCLUDE)/%s \\\n", config.java_md_.c_str());
+    fprintf(f, "  -I%s \\\n", include_.c_str());
+    bool have_oflag = (strstr(options_.c_str(), "-O") != nullptr);
+    fprintf(f, "  %s%s \\\n", have_oflag ? "" : "-O3", options_.c_str());
+    fprintf(f, "  -DHAIL_MODULE=$(MODULE)\n");
+    fprintf(f, "LIBFLAGS := -fvisibility=default %s\n", 
+      config.is_darwin_ ? "-dynamiclib -Wl,-undefined,dynamic_lookup"
+                         : "-rdynamic -shared");
+    fprintf(f, "\n");
+    // build .so from .cpp
+    fprintf(f, "$(MODULE_SO): FORCE\n");
+    fprintf(f, "\t$(CXX) $(CXXFLAGS) -o $(MODULE).o -c $(MODULE).cpp 2> $(MODULE).err\n");
+    fprintf(f, "\t$(CXX) $(CXXFLAGS) $(LIBFLAGS) -o $@ $(MODULE).o 2>> $(MODULE).err\n");
+    fprintf(f, "\n");
+    fclose(f);
+  }
+
+public:
+  bool try_to_build() {
+    // Try to create the .new file, we hold the global lock so there is no race
+    int fd = ::open(hm_lib_.c_str(), O_WRONLY|O_CREAT|O_TRUNC, 0666);
+    if (fd < 0) {
+      perror("open");
+      assert(false);
+    }
+    ::close(fd);
+    write_mak();
+    write_cpp();
+    std::stringstream ss;
+    ss << "make -B -C " << config.module_dir_ << " -f " << hm_mak_ << " 1>/dev/null";
+    // run the command synchronously, while holding the big_mutex
+    int rc = system(ss.str().c_str());
+    if (rc != 0) {
+      std::string base(config.module_dir_ + "/hm_" + key_);
+      fprintf(stderr, "makefile:\n%s", read_file_as_string(base+".mak").c_str());
+      fprintf(stderr, "errors:\n%s",   read_file_as_string(base+".err").c_str());
+    }
+    return (rc == 0);
+  }
+};
+
+NativeModule::NativeModule(
+  const char* options,
+  const char* source,
+  const char* include
+) :
+  build_state_(kInit),
+  load_state_(kInit),
+  key_(hash_two_strings(options, source)),
+  is_global_(false),
+  dlopen_handle_(nullptr),
+  lib_name_(config.get_lib_name(key_)) {
+  // Master constructor - try to get module built in local file
+  config.ensure_module_dir_exists();
+  if (file_exists(lib_name_)) {
+    build_state_ = kPass;
+  } else {
+    // The file doesn't exist, let's build it
+    ModuleBuilder builder(options, source, include, key_);
+    build_state_ = (builder.try_to_build() ? kPass : kFail);
+  }
+}
+
+NativeModule::NativeModule(
+  bool is_global,
+  const char* key,
+  ssize_t binary_size,
+  const void* binary
+) :
+  build_state_(is_global ? kPass : kInit),
+  load_state_(is_global ? kPass : kInit),
+  key_(key),
+  is_global_(is_global),
+  dlopen_handle_(nullptr),
+  lib_name_(config.get_lib_name(key_)) {
+  // Worker constructor - try to get the binary written to local file
+  if (is_global_) return;
+  int rc = 0;
+  config.ensure_module_dir_exists();
+  if (!file_exists(lib_name_)) {
+    // We hold big_mutex so there is no race
+    int fd = open(lib_name_.c_str(), O_WRONLY|O_CREAT|O_TRUNC, 0666);
+    if (fd < 0) {
+      perror("open");
+      assert(0);
```

Comment:
Using an assert here seems wrong, asserts should be internal errors.  Maybe `exit(EXIT_FAILURE)` or set build_state_ to kFail and return?

Reply:
There's additional work to be done here with error reporting, but let's save it for another PR.

In particular, failing with assert or exit on an executor is going to appear pretty mysterious to the user (or us), the only way you'll see these errors is in the executor logs which are distributed across the cluster.  Better is to throw a Java exception which Spark will propagate back to the master.

Also, I'm not sure build_state_ == kFail is particularly useful.  We're not going to get very far with a failed build, so it is probably better to fail right away and throw an exception in the JVM.

Reply:
Agree. I should try to get it to the point where errors cause the build_state_ or load_state_ to be
kFail, and then the find_LongFuncN would propagate that back up to Scala through the NativeStatus.  Where it can then be decided whether to throw an exception.  I'll see how much of that can be done quickly.